### PR TITLE
convert stage_4 of bootloader to rust

### PR DESCRIPTION
**Pull request by acheronfail**
I've been playing around trying to convert some assembly to Rust in this repository, and I've (almost) managed to convert `stage_4` to a Rust function.

The only thing that I can't figure out, is how to calculate the `memory_map_entry_count` argument? I've tried various things and maybe I don't understand something here. 🤔 

IIUC `memory_map_entry_count` is defined here in `stage_4.s`:

```asm
movzx rcx, word ptr mmap_ent
; If I understand correctly, the `movzx` instruction above is equivalent to:
; xor rcx, rcx
; mov  cx, word ptr mmap_ent
```

But, I've even tried changing that completely (to things like `mov rcx, 0x0`) and the result is always `6` when `load_elf` is called, so I'm really quite confused.

In any case, I'm putting this PR up to see if you can help me understand that part above, and to try and remove more assembly in favour of more Rust. 😄

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -63,6 +62,43 @@ impl IdentityMappedAddr {
...
+
+#[no_mangle]
+pub unsafe extern "C" fn stage_4() -> ! {
+    asm!("mov bx, 0x0" :::: "intel");
+    asm!("mov ss, bx" :::: "intel"); // Set stack segment
+
+    let kernel_start = 0x400000;
+    let kernel_size = _kib_kernel_size as u64;
+    let memory_map_addr = &_memory_map as *const _ as u64;
+    let memory_map_entry_count = 6;
```
Comment (acheronfail) [inline src/main.rs]: This is just hard-coded: from my PR description you can see I don't know how this value always resolves to `6`.

**Comment by phil-opp** [comment]
Thanks for the PR! Looks good to me overall.

> But, I've even tried changing that completely (to things like mov rcx, 0x0) and the result is always 6 when load_elf is called, so I'm really quite confused.

This seems really strange.

**Comment by acheronfail** [comment]
I've figured out why changing the instructions wasn't working - I was using `;` to comment out instructions, but as it turns out that doesn't actually comment them out. Using `#` instead does. So that basically means that the original code I thought was commented out was actually running. 🤦‍♂️ 

Knowing that, I can finish this PR - I'll convert the rest of `stage_4` to Rust and update this.

**Comment by acheronfail** [comment]
What a silly error that was!

In any case, I think this should be good to go now. 👍

File: src/main.rs [hunk 2]
```diff
@@ -63,8 +62,44 @@ impl IdentityMappedAddr {
...
+    static __page_table_end: usize;
+    static __bootloader_end: usize;
+    static __bootloader_start: usize;
+}
+
 #[no_mangle]
-pub extern "C" fn load_elf(
+pub unsafe extern "C" fn stage_4() -> ! {
+    asm!("mov bx, 0x0" :::: "intel");
+    asm!("mov ss, bx" :::: "intel"); // Set stack segment
```
Comment (phil-opp) [inline src/main.rs]: Could you merge these two into a single `asm!` statement? This ensures that they are executed right after each other. Also, we should add `bx` to the clobber list.

**Comment by phil-opp** [comment]
> I was using `;` to comment out instructions, but as it turns out that doesn't actually comment them out. Using `#` instead does. So that basically means that the original code I thought was commented out was actually running. man_facepalming

Ah yes, inline assembly is for some reasons a bit different to nasm assembly. No idea why they have different comment styles.

**Pull request merged**